Add stylelint

karlhorky opened the pull request:
Hi there! First of all, thanks to all the contributors for this awesome project. Having all the tools, configuration and setup in one place helps tons.

`create-react-app` doesn't have any CSS linting yet. I noticed that there was tentative agreement for further consideration in https://github.com/facebookincubator/create-react-app/issues/636#issuecomment-246687341, so I decided to open a PR for this, since I need it for a project anyway.

I've included [the standard config](https://github.com/stylelint/stylelint-config-standard) because I don't have a full opinionated `stylelint` config yet, but this can of course be changed.

~~**Note:** This depends on my fork of @vieron's stylelint webpack plugin, because the errors were not displaying correctly using watch mode in CRA. I've opened a pull request on the plugin here: vieron/stylelint-webpack-plugin#43~~

**Edit:** Pull request has been merged!

Olivia101 commented:
Or you could use a Node-only Sass [linter](https://github.com/sasstools/sass-lint) for both SASS and SCSS syntax!

karlhorky commented:
Stylelint also supports CSS-like syntaxes like SCSS, SugarSS and Less.

The `sasslint-webpack-plugin` is also deprecated in favor of Stylelint: https://github.com/alleyinteractive/sasslint-webpack-plugin/blob/master/README.md

karlhorky commented:
vieron/stylelint-webpack-plugin#43 has been merged and released as `0.4.1`! I've updated the dependency accordingly.

JaKXz commented on packages/react-scripts/package.json:
```diff
@@ -60,6 +61,9 @@
     "style-loader": "0.13.1",
+    "stylelint": "7.6.0",
+    "stylelint-config-standard": "15.0.0",
+    "stylelint-webpack-plugin": "0.4.1",
```
@karlhorky I recommend another update here to capture our fixes from this morning.

karlhorky replied:
Done!

gaearon commented:
Can you provide some details about node_modules disk size before and after this change please?

Also, would this work well for people compiling Sass/Less into CSS, and [importing the compiled CSS](https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/template/README.md#adding-a-css-preprocessor-sass-less-etc)?

Does this plugin work well with Webpack 2?

karlhorky commented:
> some details about node_modules disk size before and after

`du -sh node_modules` before and after (13M difference):
```sh
109M    node_modules # before
122M    node_modules # after
```

> would this work well for people compiling Sass/Less into CSS

It could lint the original source Sass/Less if I changed [the file extensions](https://github.com/facebookincubator/create-react-app/pull/1216/commits/9e851fb266e002735421d99f7914ba9f75df69a0#diff-a7f98c18479be87c9f33e7604dbd1a09R216) that are matched by the plugin. I however [only enable linting of source files](https://github.com/facebookincubator/create-react-app/pull/1216/commits/afbb7f776ddcb02bd208031d999aa61c3d7af462) (disabling the linting of any built CSS files) because the built output may not be able to be changed by the author.

> Does this plugin work well with Webpack 2?

I tested this with a simple application and the plugin works. As for working *well*, I haven't tested extensively yet - maybe @JaKXz can offer some insight here?

karlhorky commented:
I've modified the `eject` script to create a `stylelint` configuration entry as well, although JaKXz/stylelint-webpack-plugin#67 may interfere until it's resolved.

JaKXz commented:
> Does this plugin work well with Webpack 2?

The plugin is actively tested against webpack 2 as of JaKXz/stylelint-webpack-plugin#46, so I have confidence in the logic against webpack's compiler.

gaearon commented:
>It could lint the original source Sass/Less if I changed the file extensions that are matched by the plugin. I however only enable linting of source files (disabling the linting of any built CSS files) because the built output will be optimized and is thus not a good candidate for linting.

The scenario I meant is when people compile Sass to CSS but keep it in `src` next to the components. That seemed like the most straightforward way to use a preprocessor to me.

karlhorky commented:
> compile Sass to CSS but keep it in src next to the components

Ah yeah, that would also be supported by changing the file extensions.

JaKXz commented on packages/react-scripts/config/webpack.config.dev.js:
```diff
@@ -207,6 +208,14 @@ module.exports = {
+      configFile: path.join(__dirname, '../.stylelintrc'),
+      // @remove-on-eject-end
+      context: paths.appSrc,
+      files: ['**/*.css'],
```
one thing to consider is that `stylelint` can handle `.sass`/`.scss` files, and the [default glob for stylelint-webpack-plugin supports both](https://github.com/JaKXz/stylelint-webpack-plugin/blob/master/lib/constants.js#L4) so you could drop this line.

karlhorky replied:
Ah, nice idea. If @gaearon agrees we want to go this direction, I will make this change.

karlhorky commented:
@gaearon any feedback to the `files` config option? Should I just remove it as @JaKXz suggested to allow Stylelint to lint all supported file extensions?

shai32 commented:
Is this PR still active?, will it be merged?

karlhorky commented:
If there's any interest I can rebase and fix conflicts (and upgrade to webpack 2 if necessary).

roytz commented:
There is much interest :) (and to the upgrade to webpack 2 too)

cr101 commented:
Handling `.sass/.scss` files is essential

shai32 commented:
@karlhorky please upgrade to webpack 2

shai32 commented:
Is this pr active?

karlhorky commented:
I've rebased, fixed the conflicts and moved to using inline configuration instead of a `.stylelintrc` file (to follow the way that Babel and ESLint configuration is handled).

karlhorky commented:
It appears the one failing Travis test is failing also on other (unrelated) pull requests: https://travis-ci.org/facebookincubator/create-react-app/jobs/227796566

karlhorky commented:
Any further feedback @gaearon?

Timer commented:
Hey, @karlhorky! Sorry this has been on hold for so long.

I do not believe we are ready to prescribe [`stylelint-config-standard`](https://github.com/stylelint/stylelint-config-standard), but I would love to enable rules strictly for preventing errors.

> **Catch errors:** e.g. invalid hex colors, indistinguishable colors, or overriding shorthand properties.

We can consider some style-enforcing ones in the future, but they don't fit too well in our philosophy.

If you have some free time, would you mind pulling together a config strictly for finding errors?

shai32 commented:
if we remove style rule, we need a way to enable configuration to it (without eject) for users who want styling.

karlhorky commented:
> If you have some free time, would you mind pulling together a config strictly for finding errors?

Sure! I guess this could be published as another package in the monorepo.

Timer commented:
> Sure! I guess this could be published as another package in the monorepo.

Yup! Just like the babel preset and eslint config. 😄

gaearon commented:
I'll push it back to 0.11 but if we can address the feedback quickly, might still get into 0.10.
Thanks!

karlhorky commented:
Okay, I have added `stylelint-config-react-app@0.0.1` with only the stylelint rules unrelated to code style (4522aa73be1294de0dec9946e92fe72aaf9068f6). I have also added the [csstree syntax validator plugin](https://github.com/csstree/stylelint-validator) to error out on more advanced syntax errors.

karlhorky commented:
Since the tests were failing I added the stylelint dependencies to the stylelint config in 56d9ac43ab5701208853951fc0387337f8fe6a36. Not sure if this is the correct approach.

Timer commented:
> Since the tests were failing I added the stylelint dependencies to the stylelint config in 56d9ac4. Not sure if this is the correct / best approach.

Not sure if this is the correct fix... it should work similar to our eslint config.

zhaozhiming commented on packages/react-scripts/config/webpack.config.dev.js:
```diff
@@ -225,6 +226,18 @@ module.exports = {
+      },
+      configBasedir: paths.ownPath,
+      // @remove-on-eject-end
+      allowEmptyInput: true,
```
It seems `allowEmptyInput` have been removed.  
https://github.com/stylelint/stylelint/pull/2464

karlhorky replied:
Hm, right, this was merged into the `v8` branch. However version 8 is not yet out. The current version on npm is `7.10.1` however, so we cannot yet switch to this.

Timer replied:
Can we note that this should be removed when we update to ^8?

karlhorky commented:
> Not sure if this is the correct fix... it should work similar to our eslint config.

Ok, it looks like it works if we run the build script within the `react-scripts` package directory.

Would it be desirable to run it from there? I have committed this in df58946f86dfaf2ba9e128054dc501fd80b1984e. It feels a bit hacky though...

**Resolution of `stylelint-csstree-validator` not working**
```sh
➜  create-react-app git:(add-stylelint) ✗ node packages/react-scripts/scripts/build.js        
Creating an optimized production build...
Failed to compile.

Could not find "stylelint-csstree-validator". Do you need a `configBasedir`?
```

**Resolution of `stylelint-csstree-validator` working**
```sh
➜  create-react-app git:(add-stylelint) ✗ cd packages/react-scripts
➜  react-scripts git:(add-stylelint) ✗ node scripts/build.js
Creating an optimized production build...
Compiled successfully.

File sizes after gzip:

  46.75 KB  build/static/js/main.4576b8c9.js
  289 B     build/static/css/main.9a0fe4f1.css

The project was built assuming it is hosted at the server root.
To override this, specify the homepage in your package.json.
For example, add this to build it for GitHub Pages:

  "homepage" : "http://myname.github.io/myapp",

The ../../build folder is ready to be deployed.
You may serve it with a static server:

  npm install -g serve
  serve -s build
```

karlhorky commented:
Alternatively, we could set the working directory in the build.js script itself.

karlhorky commented:
Changing the directory caused more problems with other scripts unfortunately, and I wasn't able to get it working with `process.chdir()` either. This direction would probably lead to further confusion and refactoring down the road, so I opted for the simple solution of keeping `stylelint-csstree-validator` as a dev dependency of `stylelint-config-react-app`.

@Timer thoughts?

This is caused by `stylelint` calling [getModulePath()](https://github.com/stylelint/stylelint/blob/04d454664c0f2ac902367562a50648160c4e5be4/lib/utils/getModulePath.js#L7) with a `baseDir` of the directory containing the config (from [`augmentedConfig`](https://github.com/stylelint/stylelint/blob/04d454664c0f2ac902367562a50648160c4e5be4/lib/augmentConfig.js#L48)). I suppose in this constellation (calling webpack from a script within a monorepo package which uses a relative config from another monorepo package) it is expected that `stylelint-csstree-validator` will be configured as a dev dependency of the config package.

cc @davidtheclark

Timer commented:
I'll be able to take a look in a couple hours, cheers!

Timer commented:
Actually took a quick peek. Using a dev dependency is the correct solution. The original solution of a real dependency was what I was referring to as wrong.

Is this pr ready for review?

karlhorky commented:
Yep, good to go.

karlhorky commented:
@Timer yep, good to go.

karlhorky commented:
@Timer just merged in `master` and fixed the conflicts. Any feedback?

Timer commented:
@karlhorky sorry, this does LGTM!

Since we just cut 1.0.0 this can't be merged until a 2.0.0 branch is started (which is sooner than later).

Thanks for following up.

karlhorky commented:
Sounds great, thanks!

gaearon commented:
What is the impact of this on the `node_modules` size?

karlhorky commented:
@gaearon The answer I gave before to that question [is above](https://github.com/facebookincubator/create-react-app/pull/1216#issuecomment-279232164). If there's still interest in this PR I could rebase and update all dependencies and check the size again.

gaearon commented:
Lol I'm sorry 😛 
Yes, please do rebase.

karlhorky commented:
Ok, I'll get to this in the next day or so.

gaearon commented:
I'd also like to better understand the performance implications of this. If the project has e.g. 500 300-line CSS files, how much does this plugin affect the development and production build times?

People already complain about the builds being super slow so I'm not really sure we have the capacity to add any more build-time work.

JaKXz commented:
Perhaps this should be opt-in as well? 

FYI v0.10.1 of the plugin was [released recently](https://github.com/JaKXz/stylelint-webpack-plugin/releases).

gaearon commented:
>Perhaps this should be opt-in like some others in #3815?

This seems a bit trickier. We won't make those opt-in by some configuration flag. Instead if you import `.sass`, we'll ask you to install `node-sass`. If you install it, Sass will "just work". Similarly, if you import `.module.css` instead of `.css`, we will treat it as CSS Modules. But with Stylelint we can't "guess" whether you want it or not like this. We also don't know if you're *writing* CSS or compiling to it. I'm also not sure if it handle CSS Modules. So it seems like this is potentially more involved than it seems. Could the user just install and run Stylelint themselves? I'm not sure it needs to be integrated with the system in the same way that e.g. Sass needs to be for a good experience.

JaKXz commented:
I think the intention here is to have stylelint errors side by side with eslint errors in the build, and assuming CRA works with the better errors webpack plugin that should be achievable :) but I understand your point, I agree it's much harder to pinpoint exactly what the user would want from the outset, and that could change with a lot more volatility.

gaearon commented:
Retagging as 2.x because I don't think this should block the 2.0 release.
If we can get the numbers that show it's not expensive, sure, let's get it in.

If it slows down our already slow builds I don't think we can do this (but you can always use it separately in your IDE etc).

caghand commented:
Yes, you can use stylelint in your IDE, but since the errors do not appear in the build output (like eslint), you won't be able to implement a check-in policy that prevents style-violating CSS from being pushed in. :)

-Caghan

sstruct commented:
I'd prefer [Prettier](https://github.com/prettier/prettier)

oriSomething commented on packages/stylelint-config-react-app/index.js:
```diff
@@ -0,0 +1,24 @@
+  rules: {
+    'at-rule-no-unknown': true,
+    'color-no-invalid-hex': true,
+    'declaration-block-no-duplicate-properties': true,
```
Wouldn't be better to use [declaration-block-no-duplicate-properties](https://stylelint.io/user-guide/rules/declaration-block-no-duplicate-properties/) with `ignore: ["consecutive-duplicates"]`?

Moreover, is this rule blocks the pattern of (Added example)?
```
margin: 1rem;
margin-bottom: 2rem;
```

ai commented:
If we have ESLint for JS, we definitely should have Stylelint for CSS.
